### PR TITLE
fix stores without address translation

### LINKED ISSUE #196: JIT miscompiles direct stores when address translation is disabled
## Overview

With `Config::enable_address_translation=false`, the x86-64 JIT stores guest
`r0` instead of the requested value. The interpreter and JIT both return
success and consume the same number of instructions, but leave different bytes
at the target address.

Minimal program:

```asm
stdw [r1+0], 0x11223344
exit
```

```text
7a 01 00 00 44 33 22 11
95 00 00 00 00 00 00 00
```

Configuration:

```text
sBPF version:                V0
enable_address_translation: false
instruction budget:         20
```

Observed with valid writable target and decoy mappings:

```text
Interpreter target: 44 33 22 11 00 00 00 00
JIT target:         little-endian guest r0
```

## Tested Version

```text
Anza sbpf: db4f0681951171ee97988989695ceef67fe3dbb3 (v0.22.0)
Host:      Linux x86-64
Rust:      rustc 1.96.0
```

## Root Cause

The JIT saves the intended value in `[rsp-96]` and computes the target address
in `r11` ([`jit.rs`, lines 1184-1206](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/jit.rs#L1184-L1206)).
The direct-store branch then emits
[`xchg(S64, RSP, REGISTER_MAP[0], ...)`](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/jit.rs#L1236-L1244).

`X86Instruction::xchg` stores the source as the first operand and the
destination as the second
([`x86.rs`, lines 355-372](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/x86.rs#L355-L372)).
Indirect addressing uses the second operand as its memory base
([addressing definition](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/x86.rs#L73-L77),
[ModRM/SIB encoding](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/x86.rs#L118-L194)).
The generated instruction is therefore:

```text
48 87 64 20 a0    xchg rsp, [rax-96]
```

instead of:

```text
48 87 44 24 a0    xchg rax, [rsp-96]
```

[`REGISTER_MAP[0]` is `rax`](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/jit.rs#L241-L253),
which contains guest `r0`. The following store therefore writes `r0` to
`[r11]`.

Immediate stores also require reconstruction of `constant - lower_key`; the
matching `+ lower_key` currently exists only in the translated-store path
([`jit.rs`, lines 1646-1649](https://github.com/anza-xyz/sbpf/blob/db4f0681951171ee97988989695ceef67fe3dbb3/src/jit.rs#L1646-L1649)).

## Suggested Fix

Reverse both direct-store `xchg` operands, or replace the exchange sequence.
Also reconstruct sanitized immediate values in the no-translation branch and
add interpreter/JIT tests for immediate and register stores.

## Reproduction

The following standalone PoC uses live writable mappings for both the intended
target and the unintended `r0-96` access.

`src/main.rs`:

```rust
#[cfg(not(all(target_arch = "x86_64", target_os = "linux")))]
compile_error!("this reproducer requires Linux x86-64");

use libc::{
    c_void, mmap, munmap, MAP_ANONYMOUS, MAP_FAILED, MAP_PRIVATE, PROT_READ, PROT_WRITE,
};
use solana_sbpf::{
    aligned_memory::AlignedMemory,
    ebpf::{self, HOST_ALIGN},
    elf::Executable,
    memory_region::{MemoryMapping, MemoryRegion},
    program::{BuiltinProgram, FunctionRegistry, SBPFVersion},
    verifier::RequisiteVerifier,
    vm::{CallFrame, Config, ContextObject, EbpfVm, ExecutionMode},
};
use std::{ptr, ptr::NonNull, slice, sync::Arc};

const PAGE_SIZE: usize = 4096;
const BUDGET: u64 = 20;
const PROGRAM: [u8; 16] = [
    0x7a, 0x01, 0x00, 0x00, 0x44, 0x33, 0x22, 0x11, // stdw [r1], 0x11223344
    0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // exit
];

struct Page(*mut c_void);

impl Page {
    fn new() -> Self {
        let ptr = unsafe {
            mmap(
                ptr::null_mut(),
                PAGE_SIZE,
                PROT_READ | PROT_WRITE,
                MAP_PRIVATE | MAP_ANONYMOUS,
                -1,
                0,
            )
        };
        assert_ne!(ptr, MAP_FAILED, "mmap failed: {}", std::io::Error::last_os_error());
        Self(ptr)
    }

    fn address(&self) -> u64 {
        self.0.expose_provenance() as u64
    }

    fn bytes(&mut self) -> &mut [u8] {
        unsafe { slice::from_raw_parts_mut(self.0.cast(), PAGE_SIZE) }
    }
}

impl Drop for Page {
    fn drop(&mut self) {
        assert_eq!(unsafe { munmap(self.0, PAGE_SIZE) }, 0);
    }
}

struct Context {
    remaining: u64,
    memory_mapping: MemoryMapping,
}

impl ContextObject for Context {
    fn consume(&mut self, amount: u64) {
        self.remaining = self.remaining.saturating_sub(amount);
    }

    fn get_remaining(&self) -> u64 {
        self.remaining
    }

    fn active_mapping_ptr(&mut self) -> NonNull<MemoryMapping> {
        NonNull::from_mut(&mut self.memory_mapping)
    }
}

fn run(
    executable: &Executable<Context>,
    mode: ExecutionMode,
    target: &mut Page,
    decoy: &mut Page,
) -> [u8; 8] {
    target.bytes()[..8].fill(0xaa);
    decoy.bytes()[..104].fill(0);

    let target_address = target.address();
    let initial_r0 = decoy.address() + 96;
    let config = executable.get_config();
    let version = executable.get_sbpf_version();
    let mut stack = AlignedMemory::<{ HOST_ALIGN }>::zero_filled(config.stack_size());
    let mut heap = AlignedMemory::<{ HOST_ALIGN }>::with_capacity(0);
    let stack_len = stack.len();
    let regions = vec![
        executable.get_ro_region(),
        MemoryRegion::new_gapped(
            &raw mut *stack.as_slice_mut(),
            ebpf::MM_STACK_START,
            config.stack_frame_size as u64,
        ),
        MemoryRegion::new(&raw mut *heap.as_slice_mut(), ebpf::MM_HEAP_START),
        MemoryRegion::new(&raw mut target.bytes()[..PAGE_SIZE], target_address),
    ];
    let memory_mapping =
        unsafe { MemoryMapping::new(regions, config, version) }.expect("valid memory map");
    let mut context = Context {
        remaining: BUDGET,
        memory_mapping,
    };
    let mut vm = EbpfVm::new(
        executable.get_loader().clone(),
        version,
        &mut context,
        stack_len,
    );
    vm.registers[0] = initial_r0;
    vm.registers[1] = target_address;

    let label = match mode {
        ExecutionMode::Interpreted => "interpreter",
        ExecutionMode::Jit => "jit",
        ExecutionMode::PreferJit => unreachable!(),
    };
    let (count, result) = match mode {
        ExecutionMode::Interpreted => {
            let mut frames = vec![CallFrame::default(); config.max_call_depth];
            vm.execute_program(executable, &mut ExecutionMode::Interpreted, &mut frames)
        }
        ExecutionMode::Jit => vm.execute_program(executable, &mut ExecutionMode::Jit, &mut []),
        ExecutionMode::PreferJit => unreachable!(),
    };
    let final_pc = vm.registers[11];
    drop(vm);

    let mut bytes = [0; 8];
    bytes.copy_from_slice(&target.bytes()[..8]);
    assert_eq!(format!("{result:?}"), format!("Ok({initial_r0})"));
    assert_eq!(count, 2);
    assert_eq!(final_pc, 0);
    assert_eq!(context.remaining, BUDGET - 2);
    assert_eq!(&decoy.bytes()[..8], &[0; 8]);
    println!("{label}: r0=0x{initial_r0:016x} target={bytes:02x?}");
    bytes
}

fn main() {
    let version = SBPFVersion::V0;
    let config = Config {
        enable_address_translation: false,
        enabled_sbpf_versions: version..=version,
        ..Config::default()
    };
    let loader = Arc::new(BuiltinProgram::new_loader(config));
    let executable = Executable::<Context>::from_text_bytes(
        &PROGRAM,
        loader,
        version,
        FunctionRegistry::default(),
    )
    .expect("program must load");
    executable.verify::<RequisiteVerifier>().expect("program must verify");
    executable.jit_compile().expect("JIT compilation must succeed");

    let mut target = Page::new();
    let mut decoy = Page::new();
    let interpreted = run(&executable, ExecutionMode::Interpreted, &mut target, &mut decoy);
    let jitted = run(&executable, ExecutionMode::Jit, &mut target, &mut decoy);

    assert_eq!(interpreted, [0x44, 0x33, 0x22, 0x11, 0, 0, 0, 0]);
    assert_eq!(jitted, (decoy.address() + 96).to_le_bytes());
    assert_ne!(interpreted, jitted);
}
```

Example output (addresses vary):

```text
interpreter: r0=0x00007356ab3e6060 target=[44, 33, 22, 11, 00, 00, 00, 00]
jit: r0=0x00007356ab3e6060 target=[60, 60, 3e, ab, 56, 73, 00, 00]
```

Both modes should instead leave the target as:

```text
44 33 22 11 00 00 00 00
```

### PR DESCRIPTION
**Pull request by nagisa**
Fixes #196
Stacked on #198

### CODE REVIEW

**Review by Lichtso — DISMISSED** [review]

File: src/jit.rs [hunk 1]
```diff
@@ -1233,15 +1235,27 @@ impl<'a, C: ContextObject> JitCompiler<'a, C> {
...
+                        8 => self.emit_ins(X86Instruction::store(OperandSize::S64, reg, REGISTER_SCRATCH, X86IndirectAccess::Offset(0))),
+                        _ => unreachable!(),
+                    }
+                }
+                Some(Value::Constant64(val, _)) => {
+                    match len {
+                        1 => self.emit_ins(X86Instruction::store_immediate(OperandSize::S8,  REGISTER_SCRATCH, X86IndirectAccess::Offset(0), val)),
+                        2 => self.emit_ins(X86Instruction::store_immediate(OperandSize::S16, REGISTER_SCRATCH, X86IndirectAccess::Offset(0), val)),
+                        4 => self.emit_ins(X86Instruction::store_immediate(OperandSize::S32, REGISTER_SCRATCH, X86IndirectAccess::Offset(0), val)),
+                        8 => self.emit_ins(X86Instruction::store_immediate(OperandSize::S64, REGISTER_SCRATCH, X86IndirectAccess::Offset(0), val)),
```
Comment (nagisa) [inline src/jit.rs]: Hm, having had the discussion on that other PR, this particular one is probably wrong (because there isn't `mov imm64, r/m64`.) Good opportunity to write more tests for this too...

Comment (nagisa) [reply src/jit.rs]: Hm, actually could this be fine by virtue that there are no store instructions that could specify a 64-bit immediate here. The only bpf instruction that would allow such an immediate is lddw. But then maybe an assert would be good here 🤔

Comment (nagisa) [reply src/jit.rs]: I added an assert for this (and also expanded the test a bit more to use negative numbers to verify sign extension behaviour.)

**Review by LucasSte — APPROVED** [review]
